### PR TITLE
Generic/InlineControlStructure: stop listening for T_SWITCH

### PR DESCRIPTION
**Pull request by rodrigoprimo**
# Description

As originally discussed in https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/482#pullrequestreview-2062765716 (see the "Commit 2 - stop listening for T_SWITCH" section), this PR changes the `Generic.ControlStructures.InlineControlStructure` sniff to stop listening for `T_SWITCH` tokens as there is no inline version of a `switch` in [PHP](https://www.php.net/manual/en/control-structures.switch.php) and [JS](https://tc39.es/ecma262/multipage/ecmascript-language-statements-and-declarations.html).

Before this change, the sniff would bail early for a `T_SWITCH` token either because [it has a scope opener](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/47306f8a29442ac949a7b28c8d2796f89980e418/src/Standards/Generic/Sniffs/ControlStructures/InlineControlStructureSniff.php#L70-L73) or because of [a syntax error](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/47306f8a29442ac949a7b28c8d2796f89980e418/src/Standards/Generic/Sniffs/ControlStructures/InlineControlStructureSniff.php#L114-L120).

Code samples used for the sniff tests that contained a `T_SWITCH` token were updated to reflect this change. Some of those were actually added as Tokenizer tests before PHPCS had dedicated Tokenizer tests. Below there is more information about each of the modified tests answering the questions left by @jrfnl in https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/482#pullrequestreview-2062765716.

> line 82-89: related to https://github.com/squizlabs/PHP_CodeSniffer/issues/497 - looks like this test is about switch and the alternative syntax, which is no longer relevant in this sniff, but the use of ... ?> ... <?php ... within a control structure is an interesting case. Can you think of a code sample which would maintain the value of this test, but uses one of the other control structures ?
Also: should the original test - which was really a tokenizer bug - be moved to a Tokenizer test ?

I updated the original code sample used in the sniff test to use `T_IF` and `T_ELSEIF` instead of `T_SWITCH`. I also created a new Tokenizer::recurseScopeMap() test file and added tests to ensure that the scope indexes are set correctly for switches using the normal and alternative syntaxes.

> line 93 - 96: also related to https://github.com/squizlabs/PHP_CodeSniffer/issues/497 - looks like this test is about switch and the alternative syntax, which is no longer relevant in this sniff, but the use of nested alternative syntax control structures could still be a valid test case. Can you think of a code sample which would maintain the value of this test, but uses one of the other control structures ?
Also: should the original test - which was really a tokenizer bug - be moved to a Tokenizer test ?

Similar to the above, I updated the code sample to use `T_FOREACH` instead of `T_SWITCH` and created dedicated Tokenizer tests to check setting the scope for `T_IF` with a nested `T_SWITCH` and `T_CASE` when `T_CASE` is missing `T_BREAK`.

> Line 109 - 117: related to https://github.com/squizlabs/PHP_CodeSniffer/issues/543 - looks like this test is about a closure being used within the condition of a switch, which is no longer relevant in this sniff, but the use of the closure in the condition is still an interesting case. Can you think of a code sample which would maintain the value of this test, but uses one of the other control structures (without curlies) ?
Also: should the original test - which was really a tokenizer bug - be moved to a Tokenizer test ?

I replaced the code sample with an inline `T_IF` with a closure within the condition. Also, I added a dedicated Tokenizer test to ensure that the scope for a `T_SWITCH` is set correctly when there is a closure within the condition.

> Line 146 - 155: related to https://github.com/squizlabs/PHP_CodeSniffer/issues/879 - looks like this test is about mixing if/elseif/else with braces and without braces in one code sample. I imagine, this test is still valid as-is, or maybe the switch control structure wrapper should be replaced with another control structure wrapper ?
Also: should the original test - which was, in part, a tokenizer bug - be moved to a Tokenizer test ?

I left the code sample as is because I believe it is still valid. The original problem only manifested when the `T_IF` was nested within a `T_SWITCH`. A dedicated tokenizer test was added using the same code sample.

> Line 204 - 213: related to https://github.com/squizlabs/PHP_CodeSniffer/issues/1590 - looks like this test is about adding braces to an if that's returning a nested function call. I imagine, this test is still valid as-is, or maybe the switch control structure wrapper should be replaced with another control structure wrapper ?
Also: should the original test - which was, in part, a tokenizer bug - be moved to a Tokenizer test ?

I replaced the `T_SWITCH` in the original code sample with a `T_FOR` and created a dedicated Tokenizer test.

> Test case file *.js:
    line 23 - can a meaningful replacement test be created which would use the keyword of one of the other control structures ?

I did replace `switch` with `for`. But it is important to note that the original code sample lost part of its value over time as now the sniff bails early if there is no opening parenthesis after the control structure. Ideally, it should be moved to a tokenizer test, but since there are no tests for the JS tokenizer, and support will be dropped in
PHPCS 4.0, I opted to simply use a control structure other than `T_SWITCH`. This test was originally added in https://github.com/PHPCSStandards/PHP_CodeSniffer/commit/b3f4f83b2c1ad1a07086a8af09cb371fb0f4b1a4.

## Suggested changelog entry

`Generic.ControlStructures.InlineControlStructure` stop listening for T_SWITCH as there is no inline version of this control structure in PHP and JS

## Related issues/external references

First discussed in https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/482#pullrequestreview-2062765716

## Types of changes

- [ ] Bug fix _(non-breaking change which fixes an issue)_
- [ ] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [ ] Documentation improvement

## PR checklist

- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [x] I have added tests to cover my changes.
- [x] I have verified that the code complies with the projects coding standards.
- [ ] \[Required for new sniffs\] I have added XML documentation for the sniff.

### CODE REVIEW

**Comment by rodrigoprimo** [comment]
@jrfnl, as we discussed last week, I reorganized the commits in this PR so that the changes to the sniff tests are not all in the commit that changes the InlineControlStructure sniff to stop listening for `T_SWITCH`. Instead, each change to the sniff tests is in the commit that creates the related tokenizer test.

This PR is now ready to be reviewed.

File: tests/Core/Tokenizers/Tokenizer/RecurseScopeMapSwitchTokenScopeTest.inc
Comment (jrfnl) [inline tests/Core/Tokenizers/Tokenizer/RecurseScopeMapSwitchTokenScopeTest.inc:28]: The test could be made even more "interesting" if the closure had a return type declaration (which also uses the `T_COLON` token).

Might actually even be better to have a second test sample for that.

On that note, maybe we should also have a test with an arrow function callback with/without return type declaration in a switch condition ?

**Note:** the above remarks do not need to be addressed in this PR. They can be left for a follow-up PR to further improve the tokenizer tests.

File: tests/Core/Tokenizers/Tokenizer/RecurseScopeMapSwitchTokenScopeTest.inc
Comment (jrfnl) [inline tests/Core/Tokenizers/Tokenizer/RecurseScopeMapSwitchTokenScopeTest.inc:30]: Nitpick: is this marker name clear enough ? Have you considered a name like `testSwitchClosureWithinConditionScopeOpener` to tie the test marker to the code sample and reduce the risk of duplicate marker names in test case files ?

File: tests/Core/Tokenizers/Tokenizer/RecurseScopeMapSwitchTokenScopeTest.inc
Comment (jrfnl) [inline tests/Core/Tokenizers/Tokenizer/RecurseScopeMapSwitchTokenScopeTest.inc:16]: Suggestion: `testSwitchAlternativeSyntaxEnsureTestWillNotPickUpWrongColon`.

I realize that makes it a very long name, but it will reduce the chance of there being duplicate markers once more tests get added to the file and it closely ties the marker name to the code sample test marker name, making it clear these test markers belong together and are related to the same code sample.

File: tests/Core/Tokenizers/Tokenizer/RecurseScopeMapCaseKeywordConditionsTest.inc
Comment (jrfnl) [inline tests/Core/Tokenizers/Tokenizer/RecurseScopeMapCaseKeywordConditionsTest.inc:1]: Should this improvement also be done `RecurseScopeMapDefaultKeywordConditionsTest` tests ?

File: tests/Core/Tokenizers/Tokenizer/RecurseScopeMapCaseKeywordConditionsTest.inc
Comment (jrfnl) [inline tests/Core/Tokenizers/Tokenizer/RecurseScopeMapCaseKeywordConditionsTest.inc:107]: Would it make sense to have a similar test to this one in the `RecurseScopeMapDefaultKeywordConditionsTest` test ?

**Review by jrfnl — COMMENTED** [review]
@rodrigoprimo Thanks for this PR and sorry it took me so long to get back to this one.

I've gone through the code now and left some inline remarks.
Not everything needs to be addressed in this PR. It is perfectly fine to open an issue that a follow-up will be needed to the PR at a later point in time to further stabilize and expand the Tokenizer tests.

All in all, looking good and thank you so much for reorganizing the commits. Made it much more straight forward to review now.

File: tests/Core/Tokenizers/Tokenizer/RecurseScopeMapSwitchTokenScopeTest.php
Comment (rodrigoprimo) [reply tests/Core/Tokenizers/Tokenizer/RecurseScopeMapSwitchTokenScopeTest.php]: Done

File: tests/Core/Tokenizers/Tokenizer/RecurseScopeMapSwitchTokenScopeTest.php
Comment (rodrigoprimo) [reply tests/Core/Tokenizers/Tokenizer/RecurseScopeMapSwitchTokenScopeTest.php]: I normally try to do that, but I missed it for this one. Thanks for pointing it out. It should be fixed now.

File: tests/Core/Tokenizers/Tokenizer/RecurseScopeMapIfKeywordConditionsTest.inc
Comment (rodrigoprimo) [reply tests/Core/Tokenizers/Tokenizer/RecurseScopeMapIfKeywordConditionsTest.inc:4]: AFAICS, it is not necessary. Even with a single `if`, the same problem (mismatch of statements in the tokenizer causing the `if` to have no `scope_condition`, `scope_opener` and `scope_closer`) happens if the change made in fddc61a is reverted.

File: tests/Core/Tokenizers/Tokenizer/RecurseScopeMapSwitchTokenScopeTest.inc
Comment (rodrigoprimo) [reply tests/Core/Tokenizers/Tokenizer/RecurseScopeMapSwitchTokenScopeTest.inc]: Done

File: src/Standards/Generic/Tests/ControlStructures/InlineControlStructureUnitTest.1.inc
Comment (rodrigoprimo) [reply src/Standards/Generic/Tests/ControlStructures/InlineControlStructureUnitTest.1.inc]: That makes sense, I added the following: `echo !is_null($element) ? $element : 'default';`

File: tests/Core/Tokenizers/Tokenizer/RecurseScopeMapIfKeywordConditionsTest.php
Comment (rodrigoprimo) [reply tests/Core/Tokenizers/Tokenizer/RecurseScopeMapIfKeywordConditionsTest.php]: Done

File: tests/Core/Tokenizers/Tokenizer/RecurseScopeMapSwitchTokenScopeTest.inc
Comment (rodrigoprimo) [reply tests/Core/Tokenizers/Tokenizer/RecurseScopeMapSwitchTokenScopeTest.inc]: Done

**Comment by rodrigoprimo** [comment]
Thanks for your review, @jrfnl!

I answered all the inline comments I believe should be addressed in this PR. As you suggested, I plan to create a follow-up PR to address the comments that were not blocking this PR and that I didn't answer. If I cannot create the follow-up PR right after this one is merged, I will create an issue documenting what needs to be done.

File: src/Standards/Generic/Tests/ControlStructures/InlineControlStructureUnitTest.1.inc
Comment (jrfnl) [reply src/Standards/Generic/Tests/ControlStructures/InlineControlStructureUnitTest.1.inc]: Appreciated!

File: src/Standards/Generic/Tests/ControlStructures/InlineControlStructureUnitTest.1.inc
Comment (jrfnl) [reply src/Standards/Generic/Tests/ControlStructures/InlineControlStructureUnitTest.1.inc]: Ah, but the `:` in that statement would be tokenized as `T_INLINE_ELSE`, not as `T_COLON`. Want to have another look at this ?

**Comment by jrfnl** [comment]
> As you suggested, I plan to create a follow-up PR to address the comments that were not blocking this PR and that I didn't answer. If I cannot create the follow-up PR right after this one is merged, I will create an issue documenting what needs to be done.

Thanks.

> I split the new changes per related original commit, hoping this will help you combine the commits before merging. Please let me know if this approach makes sense or not.

It does 👍🏻 

> This PR is ready for another review.

@rodrigoprimo Just one comment left to address: https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/595#discussion_r1893235283

File: src/Standards/Generic/Tests/ControlStructures/InlineControlStructureUnitTest.1.inc
Comment (rodrigoprimo) [reply src/Standards/Generic/Tests/ControlStructures/InlineControlStructureUnitTest.1.inc]: I misunderstood your suggestion during our call and should have read the comment more carefully.

I just pushed a new commit updating the code to use a `T_COLON` part of a function's return type declaration:

`echo (function($element): string { return $element; } )($element);`

The code is a bit weird in that it could be simplified to just `echo $element`, but that is probably ok for this test as it ensures there is a `T_COLON` token not related to an inline control structure alternative syntax.

**Comment by rodrigoprimo** [comment]
> @rodrigoprimo Just one comment left to address: https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/595#discussion_r1893235283

@jrfnl, I just pushed a new commit addressing this comment.

File: src/Standards/Generic/Tests/ControlStructures/InlineControlStructureUnitTest.1.inc
Comment (jrfnl) [reply src/Standards/Generic/Tests/ControlStructures/InlineControlStructureUnitTest.1.inc]: Convoluted code samples are a good thing™ in the context of tests for PHPCS, so 👍🏻

**Review by jrfnl — APPROVED** [review]
Thanks @rodrigoprimo ! All good and very happy to get this one merged and out of the way.

I'm going to rebase & squash a few commits now. Merging once the build has passed (again).

File: tests/Core/Tokenizers/Tokenizer/RecurseScopeMapCaseKeywordConditionsTest.inc
Comment (rodrigoprimo) [reply tests/Core/Tokenizers/Tokenizer/RecurseScopeMapCaseKeywordConditionsTest.inc:1]: @jrfnl, it seems to me that adding the closer markers to `RecurseScopeMapDefaultKeywordConditionsTest` is not necessary as the tests there are different. The code doesn't look for the `scope_closer` using `AbstractTokenizerTestCase::getTargetToken()` and the test marker. Instead, a `closerOffset` is passed, and it is used to determine the `scope_closer`:

https://github.com/rodrigoprimo/PHP_CodeSniffer/blob/82fae8e57e10f6a7000daf53910394d109064736/tests/Core/Tokenizers/Tokenizer/RecurseScopeMapDefaultKeywordConditionsTest.php#L130

File: tests/Core/Tokenizers/Tokenizer/RecurseScopeMapSwitchTokenScopeTest.inc
Comment (rodrigoprimo) [reply tests/Core/Tokenizers/Tokenizer/RecurseScopeMapSwitchTokenScopeTest.inc:28]: I just created a PR to add the tests that you suggested: https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/812

File: tests/Core/Tokenizers/Tokenizer/RecurseScopeMapCaseKeywordConditionsTest.inc
Comment (rodrigoprimo) [reply tests/Core/Tokenizers/Tokenizer/RecurseScopeMapCaseKeywordConditionsTest.inc:100]: Here is the PR that adds a variation of this test: https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/813

File: tests/Core/Tokenizers/Tokenizer/RecurseScopeMapCaseKeywordConditionsTest.inc
Comment (rodrigoprimo) [reply tests/Core/Tokenizers/Tokenizer/RecurseScopeMapCaseKeywordConditionsTest.inc:107]: Here is the PR that adds a variation of this test: https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/813

**Pull request merged**
